webapi/src/main/scala/org/knora/webapi/messages/util/search/gravsearch/prequery/InferenceOptimizationService.scala

```diff
@@ -120,6 +122,20 @@ final case class InferenceOptimizationService(
         }
     } yield relevantOntologiesMaybe
   }
+
+  /**
+   * Returns the set of ontologies of a given project to which the search is limited.
+   *
+   * @param projectIri the IRI of the project.
+   * @return the set of ontology IRIs of the project.
+   */
+  def getProjectOntologies(projectIri: ProjectIri): Task[Option[Set[SmartIri]]] =
```

Comment:
Rather than manually having to traverse the Ontology data structure I would recommend to directly use the `OntologyRepo` in the `SearchResponder`.

Something like is much easier to read and understand:

```
val ontologyIris: List[SmartIri] = ontologyRepo.findByProject(projectIri).map(_.ontologyMetadata.ontologyIri)
val ontologyIrisAsOptionSet: Option[Set[SmartIri]] = Option.when(ontologyIris.nonEmpty)(ontologyIris.toSet)
```

Reply:
Also the `InferenceOptimizationService` should probably not have intricate knowledge about the cache data structure.

Reply:
You're right, much nicer: https://github.com/dasch-swiss/dsp-api/pull/3557/commits/5558d8e7da91ce13a0fb935b5e1e2efc24eca2bc

In other places, `InferenceOptimizationService` has way too intricate knowledge about the cache, but I would prefer to keep that out of the scope of this PR.

Reply:
ah... way, now I actually understand you suggestion, I think - will adjust the code some more.

Reply:
https://github.com/dasch-swiss/dsp-api/pull/3557/commits/d9506f307f0aa7807eb976b9d8ef4a8e4e0245d9